File: backend/src/handlers/getBookings.ts
/**
 * getBookings.ts
 * ---------------------------------------------------------------------------
 * GET /bookings?slotId=<id>
 * Returns bookings, optionally filtered by slotId (queries the GSI when given).
 *
 * SECURITY: This endpoint returns patient PII (name, phone, email) and is
 * therefore admin-only. It requires a valid `x-api-key` header matching the
 * `ADMIN_API_KEY` environment variable.
 * ---------------------------------------------------------------------------
 */

import type { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { QueryCommand, ScanCommand } from '@aws-sdk/lib-dynamodb';
import { jsonResponse, errorResponse } from '../utils/response.js';
import { dynamoClient } from '../utils/dynamo.js';
import { isAuthorized } from '../utils/auth.js';
import { z } from 'zod';

/** Zod schema validating the optional `slotId` query parameter. */
const querySchema = z.object({
  slotId: z.string().regex(/^[\w-]{1,64}$/).optional(),
});

/**
 * Lambda handler for retrieving bookings.
 *
 * @param event - API Gateway proxy event.
 * @returns 200 with a list of bookings, 400 on invalid query params,
 *          401 if unauthenticated, or 500 on unexpected error.
 */
export async function lambdaHandler(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
  // --- Authorization gate: admin-only endpoint (returns PII) ---------------
  if (!isAuthorized(event)) {
    return errorResponse(401, 'Unauthorized', undefined, event);
  }

  const bookingsTable = process.env.BOOKINGS_TABLE;
  if (!bookingsTable) {
    return errorResponse(500, 'Missing BOOKINGS_TABLE configuration', undefined, event);
  }

  try {
    const queryParams = querySchema.parse({ slotId: event.queryStringParameters?.slotId });
    const items: Array<Record<string, unknown>> = [];

    if (queryParams.slotId && queryParams.slotId !== 'all') {
      // Query the GSI by slotId for a focused result set.
      const response = await dynamoClient.send(
        new QueryCommand({
          TableName: bookingsTable,
          IndexName: 'slotId-index',
          KeyConditionExpression: 'slotId = :slotId',
          ExpressionAttributeValues: { ':slotId': queryParams.slotId },
        }),
      );
      items.push(...(response.Items ?? []));
    } else {
      // Paginate through the full table (acceptable at capstone scale).
      let exclusiveStartKey: Record<string, unknown> | undefined;
      do {
        const response = await dynamoClient.send(
          new ScanCommand({ TableName: bookingsTable, ExclusiveStartKey: exclusiveStartKey }),
        );
        items.push(...(response.Items ?? []));
        exclusiveStartKey = response.LastEvaluatedKey;
      } while (exclusiveStartKey);
    }

    // Sort newest bookings first.
    items.sort((a, b) => {
      const aBooked = Number(a.bookedAt ?? 0);
      const bBooked = Number(b.bookedAt ?? 0);
      return bBooked - aBooked;
    });

    return jsonResponse(200, items, event);
  } catch (error) {
    if (error instanceof z.ZodError) {
      return errorResponse(400, 'Invalid query parameters', { issues: error.issues }, event);
    }
    // Do NOT leak internal error details to the client.
    return errorResponse(500, 'Failed to fetch bookings', undefined, event);
  }
}

File: backend/local_handlers/get_bookings.py
"""
get_bookings.py
---------------------------------------------------------------------------
GET /bookings?slotId=<id>
Returns bookings, optionally filtered by slotId (queries the GSI when given).

SECURITY: This endpoint returns patient PII (name, phone, email) and is
therefore admin-only. It requires a valid `x-api-key` header matching the
`ADMIN_API_KEY` environment variable.
---------------------------------------------------------------------------
"""
import hmac
import json
import os
import re
import boto3
from boto3.dynamodb.conditions import Key
from decimal import Decimal

dynamodb = boto3.resource("dynamodb")
bookings_table = dynamodb.Table(os.environ["BOOKINGS_TABLE"])

CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type,Authorization,x-api-key",
    "Access-Control-Allow-Methods": "GET,OPTIONS"
}


def decimal_default(obj):
    if isinstance(obj, Decimal):
        return int(obj) if obj % 1 == 0 else float(obj)
    raise TypeError


def is_authorized(event):
    """Verify the request carries a valid admin API key.

    Uses hmac.compare_digest for constant-time comparison to mitigate
    timing attacks. Fails closed (denies) if ADMIN_API_KEY is not set.
    """
    expected = os.environ.get("ADMIN_API_KEY", "")
    if not expected:
        return False
    headers = event.get("headers") or {}
    provided = headers.get("x-api-key") or headers.get("X-Api-Key") or ""
    if not provided:
        return False
    return hmac.compare_digest(provided, expected)


def lambda_handler(event, context):
    # --- Authorization gate: admin-only endpoint (returns PII) ---------------
    if not is_authorized(event):
        return {
            "statusCode": 401,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": "Unauthorized"})
        }

    try:
        params = event.get("queryStringParameters") or {}
        slot_id = params.get("slotId")

        if slot_id and slot_id != "all":
            if not re.match(r'^[\w\-]{1,64}$', slot_id):
                return {
                    "statusCode": 400,
                    "headers": CORS_HEADERS,
                    "body": json.dumps({"error": "Invalid slotId parameter."})
                }
            response = bookings_table.query(
                IndexName="slotId-index",
                KeyConditionExpression=Key("slotId").eq(slot_id)
            )
            items = response.get("Items", [])
        else:
            response = bookings_table.scan()
            items = response.get("Items", [])
            while "LastEvaluatedKey" in response:
                response = bookings_table.scan(ExclusiveStartKey=response["LastEvaluatedKey"])
                items.extend(response.get("Items", []))

        items.sort(key=lambda b: b.get("bookedAt", ""), reverse=True)

        return {
            "statusCode": 200,
            "headers": CORS_HEADERS,
            "body": json.dumps(items, default=decimal_default)
        }
    except Exception:
        # Do NOT leak internal error details to the client.
        return {
            "statusCode": 500,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": "Failed to fetch bookings"})
        }